pkg/cluster/resources.go

```diff
@@ -4,17 +4,108 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"reflect"
 	"time"
 
 	corev1 "k8s.io/api/core/v1"
 	rbacv1 "k8s.io/api/rbac/v1"
 	k8serr "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	dscv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/datasciencecluster/v1"
+	dsciv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/dscinitialization/v1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/resources"
 )
 
+// GetSingleton retrieves a singleton instance of a Kubernetes resource of type T.
+//
+// It ensures that only one instance exists and updates the provided object pointer
+// with the retrieved data and:
+//   - If no instances are found, it returns a "NotFound" error.
+//   - If multiple instances are found, it returns an error indicating an unexpected
+//     number of instances.
+//   - A generic error in case of other failures
+//
+// Generic Parameters:
+//   - T: A Kubernetes API resource that implements client.Object.
+//     T **must be a pointer to a struct**, allowing the function to update its contents.
+//
+// Parameters:
+//   - ctx: The context for the API request, allowing for cancellation and timeouts.
+//   - cli: The Kubernetes client used to interact with the cluster.
+//   - obj: A **pointer to a struct** that implements client.Object, which will be populated with the retrieved resource.
+//
+// Returns:
+//   - nil if exactly one instance of the resource is found and successfully assigned to obj.
+//   - An error if no instances or multiple instances are found, or if any failure occurs.
+func GetSingleton[T client.Object](ctx context.Context, cli client.Client, obj T) error {
+	if reflect.ValueOf(obj).IsNil() {
+		return errors.New("obj must be a pointer")
+	}
+
+	if err := resources.EnsureGroupVersionKind(cli.Scheme(), obj); err != nil {
+		return err
+	}
+
+	objGVK := obj.GetObjectKind().GroupVersionKind()
```

Comment:
I'm not fully understanding (checked the call chain), doesn't  apiutil.GVKForObject give gvk directly?

Reply:
oh, yes. I'll amend it.

Reply:
Ended up using `resources.GetGroupVersionKindForObject` which enhances `apiutil.GVKForObject`